### ISSUE TITLE
Manually run a TPS benchmark

### ISSUE THREAD

**Issue by jakmeier** [near/nearcore#9149]
Manually run a 5M + 15M benchmark and report the results. (5M daily transaction to SocialDB, 15M other daily transactions)

This has to be with a proper state and with multi-machine infrastructure.

Progress on the set up:
- [x] generate social DB + some other traffic
- [x] use large state (e.g. testnet snapshot)
- [x] use multiple RPC nodes running on different machines (one RPC node per shard seems like a good balance)
- [x] use multiple validators running on different machines (not 100 to save costs, but ~~maybe at least 4 or 5 validators with equal stake~~ edit: I went with only 2 validators for an initial evaluation)

Progress on the results:
- [x] Collect data for first full run
- [x] Analyze and report results from first full run

Once all of this is done, we can think about necessary automation steps and what we need to improve in the setup.

**Comment by jakmeier** [comment]
By now, I managed to manually start a 2-machine network with 1 BP on one machine and one RPC node (no BP or CP) on another machine.

I used the forking script to create a new chain, then cloned the data disk to another gcp machine, and updated a few values on the new machine:
- in `config.json`
  - set the other node as boot node `ed25599:abcdef...abcdef@public-ip:port` (only on RPC node)
  - set the `min_num_peers` to 1 (both machines)
- make sure a different `node_key.json` and `validator_key.json` is used (I used an implicit account without funds, works because this is not going to be a validator)

I will shut this system down for now and do an attempt with 4+ RPC node machines tomorrow. This should give the first results with both representative state (testnet) and representative hardware for validators. I will postpone increasing the number of validators for now.

**Comment by robin-near** [comment]
A side comment - I hope we'll soon be able to expand the forking tool to support a config for multiple machines, some validators and some RPC nodes.

**Comment by jakmeier** [comment]
With some tinkering, I have been able to run a benchmark on a testnet fork with 2 validator nodes, 5 RPC nodes, and some extra load generating machines that don't even have a neard image.

I'll write a more detailed update and specific steps I had to do in later comments. For now just the summary:
- At ~280 TPS (50% FT transfer + 25% Social DB follow + 25% Social DB post) the capacity is reached. That's about 25M daily transaction, hence achieves the initial goal of 5M Social transactions + 15M other transactions.
- Because of our shard setup, it only used 2 shards. (needs improving on the setup side but also the aurora shard is a real limit)
- Because of compute costs, I'm not getting 1000 Tgas per chunk

I will let it run at around the capacity over the weekend to see if it can sustain that kind of load, or if maybe RocksDB will slow down over time.

Data so far:
- [Grafana](https://nearinc.grafana.net/d/1dZGhpJ4k/blockchain-utilization?orgId=1&var-chain_id=testnet-experimental&var-role=All&var-node_type=All&var-node_id=All&var-shard_id=3&from=1686315446376&to=1686320001827)
- ![image](https://github.com/near/nearcore/assets/6342444/ed090298-f783-4e70-8440-fd2a6abcfb36)
- ![image](https://github.com/near/nearcore/assets/6342444/475a487b-04b5-485a-876d-8d4f5a9d86da)

**Comment by jakmeier** [comment]
For reference, here are the steps I used to setup my first manual benchmarks on a testnet fork:

cc / FYI @marcelo-gonzalez @posvyatokum @robin-near 

## Prepare an image
1. Run a single testnet RPC node, make sure it is in sync and finishes data migration if necessary.
2. Stop neard.
3. Checkout https://github.com/near/nearcore/pull/9121 and update the code to include a second validator. Make sure to use different public keys and validator account ids. (I hardcoded it, [see this code snippet](https://github.com/jakmeier/nearcore/blob/78f5d6d70dead8486bbd85782158fa1edcf175a4/tools/fork-network/src/cli.rs#LL162C0-L180C10). Hopefully we will have a nicer way to do it eventually). 
4. compile it with `cargo build -r --bin neard`
5. `cp ~/nearcore/target/release/neard ~/`
6. `~/neard --unsafe-fast-startup fork-network`
7. stop the machine
8. go to gcloud and make a snapshot of the data disk

## Start the forked network
1. Create as many nodes as desired, each with their on machine and a new data disk created from the snapshot
2. Make sure neard is stopped on all machines (systemd may start it automatically depending on your image)
3. Check the external IP of one of the validator nodes. Also lookup the peer id of the node (ed25519 public key in ~/.near/node_key.json).
4. Prepare a config file that has `"boot_nodes": "ed25519:abcd....abcd@12.34.256.111:24567"` using the peer id and IP from the previous step. Also ensure `min_num_peers` is an appropriate value. For example, if you run with 3 validators and 1 RPC nodes, the default `"min_num_peers": 5` is too high and the network would never start.
5. Copy the config file to all nodes at `.near/config.json`
6. Ensure all nodes have the same neard (ideally compile it on one machine and copy it to all others)
7. Ensure all nodes have unique node keys and validators have unique validator account ids
8. run `~/neard --unsafe-fast-startup run` on all machines

## Start the load generator
1. Create new gcp instances. Ideally with more CPU cores and less memory than typical neard steup. A single 128GB disk is more than enough.
2. Firewall setting: TCP ports 8089 and 5557 should be opened. 8089 for UI access, 5557 to connect with maser node with worker nodes. (our gcp project near-core has the `locust-load-generator` tag to open these two ports)
3. Checkout nearcore master, install build essentials, python requirements, and locust
  - `sudo apt-get install build-essential python3-dev gcc`
  - `wget https://bootstrap.pypa.io/get-pip.py`
  - `sudo python3 get-pip.py`
  - `pip3 install pyopenssl --upgrade`
  - `git clone https://github.com/near/nearcore.git`
  - `pip3 install -r ./nearcore/pytest/requirements.txt`
  - `pip3 install locust`
4. Now start a locust master instance and select one of the RPC nodes IP as host
5. Wait for the master to initialize, you should be able to load the UI in the browser using HTTP at port 3030
6. Start worker instance, specifying the master IP to the machine running master node and pick the host IP as one from the RPC nodes (can be mixed to distribute load across RPC ndoes)
7. In the UI, select how many users to run a test with.

## Reset (hacky and not recommened, but for reference, this worked for me)
1. Stop the network
2. run `~/neard --unsafe-fast-startup fork-network --reset` on all nodes to reset RocksDB back to the point before the fork
3. the step above always fails with an error about some missing files, fix that last step by manually copying the old genesis.json (`cp .near/genesis.json.backup .near/data/genesis.json`)
4. (very hacky) Use the same `validator_key.json` on all machines as was used for the initial fork (but keep a backup of the real validator keys)
5. Now run `~/neard --unsafe-fast-startup fork-network` on all machines, which should give you identical genesis.json files. (Check using `shasum`, if they don't match, you probably messed up step 4. You can try to just copy the genesis.json from another machine but probably the state column in RocksDB is incomplete and will result in crashes later if not fixed. Best to start over at step 2.) 
6. Use the proper validator keys again
7. start all nodes again

**Comment by marcelo-gonzalez** [comment]
Running it on the mocknet test for 1.35.0 (kind of a similar set up as the testnet fork above), I got a similar limit for the number of txs (slightly smaller, around 220 tx/s), but the tgas was well below 1000 in each of the three shards. I checked with tcpdump to make sure it was distributing load across the different RPC nodes like I asked it to by giving different hosts on the command line for the different workers, and I only saw load to the master's host. So I'm gonna try and investigate whether I just did something wrong, or whether it needs to be changed somehow, and I might have been hitting some sort of broadcast TX RPC bottleneck when all the RPC requests end up on the same node.

**Comment by jakmeier** [comment]
@marcelo-gonzalez I have seen similar results as I have been testing your shard distribution. However, it looks like the compute cost limit is actually hit, it's just that this is hit much quicker than the gas limit. So I think it probably works as intended, we just hit the limits much faster than we hoped.

If you are running with a recent neard binary, you should get metrics in this dashboard: https://nearinc.grafana.net/d/1dZGhpJ4k/blockchain-utilization?orgId=1&refresh=30m&var-chain_id=testnet-experimental&var-role=All&var-node_type=All&var-node_id=All&var-shard_id=3&from=now-3h&to=now

In particular, the histograms for gas & copute cost per chunk shows if a chunk is full because of compute costs. For example here:
![image](https://github.com/near/nearcore/assets/6342444/94303a25-7c4b-4a2e-9a01-4a58d5021110)

**Comment by jakmeier** [comment]
Closing the loop here:

## Setup
- I ran with 2 validator machines, 5 RPC node machines, 2 locust machines with increased number of cores
- This is on top of a testnet fork
- Load was 2000 users across 25 workers, by number of in-flight requests: 
  - 60.0% ft_transfer
  - 15.0% social/follow
  - 15.0% social/post
  - 5.0% congestion/compute_sha256
  - 5.0% congestion/compute_sum
- Accounts were generated to spread the load between all shards
   - shard 1 (aurora) was not used because we can't generate new accounts for it
   - socialDB is on a single account, hence on a single shard, shard 3
   - the congestion workload is also on a single account, specifically shard 2

## Observations and Results
### Overall stable

All chunks were produced in <800ms, eventhough the compute costs per second per shard were up to 1.5s.
![image](https://github.com/near/nearcore/assets/6342444/549ef67c-9dc3-4c64-b590-3f997721c3c9)
![image](https://github.com/near/nearcore/assets/6342444/a9b18017-d7cb-40d1-8d77-43c32e1ed42d)

### Single shard congestion

The congestion workload clogs up shard 2 quickly.
We observe up to 1000 delayed receipts queued up and all processed receipts coming from the delay-queue rather than from incoming receipts directly.

Occasionally, the queue goes down to zero. That's unexpected and I can't explain it at the moment.

In case you wonder why we see 1500 Tgas, it's because the graph reports data per second, not per chunk.

![image](https://github.com/near/nearcore/assets/6342444/76993c4d-6a59-47d6-a5e5-196d3d515587)
![image](https://github.com/near/nearcore/assets/6342444/c676600a-6dcf-44c7-bb84-d5ce7e187513)

### User Experience
The throughput peaked around 170 TPS. This is lower than previous experiment because it includes heavy receipts from the congestion workload, and we don't utilize an entire shard.

![Screenshot 2023-06-21 at 18-09-42 Locust for None](https://github.com/near/nearcore/assets/6342444/d29620ca-ae83-4891-a23d-0a89a1302192)

In this experiment, the wait time for users was significantly worse for congestion load than for social DB. (19.3s average for congestion load, 10.4s average for social DB user)
That's great, it shows that we have some isolation between shards.
But even 10s wait time is not ideal

![image](https://github.com/near/nearcore/assets/6342444/cee8988a-6b98-43c9-b376-8e0093dea7c9)
